os/drivers/audio/alc1019.c

```diff
@@ -0,0 +1,1083 @@
+#endif
+        .allocbuffer = NULL,                    /* allocbuffer    */
+        .freebuffer = NULL,                     /* freebuffer     */
+        .enqueuebuffer = alc1019_enqueuebuffer, /* enqueue_buffer */
+        .cancelbuffer = alc1019_cancelbuffer,   /* cancel_buffer  */
+        .ioctl = alc1019_ioctl,                 /* ioctl          */
+        .read = NULL,                           /* read           */
+        .read = NULL,                           /* write          */
+        .reserve = alc1019_reserve,             /* reserve        */
+        .release = alc1019_release,             /* release        */
+};
+
+static int alc1019_writereg_3byte(FAR struct alc1019_dev_s *priv, uint8_t regaddr_h, uint8_t regaddr_l , uint8_t regval)
+{
+        int ret;
+        uint8_t reg[3];
+        FAR struct i2c_dev_s *dev = priv->i2c;
+        FAR struct i2c_config_s *alc1019_i2c_config = &(priv->lower->i2c_config);
+
+        reg[0] = regaddr_h;
+        reg[1] = regaddr_l;
+        reg[2] = regval;
+
+        ret = i2c_write(dev, alc1019_i2c_config, (uint8_t *)reg, 3);
+        if (ret < 0) {
+                auddbg("Error, cannot write reg %x\n", regaddr);
+        }
+        return ret;
+}
+
+/************************************************************************************
+ * Name: alc1019_exec_i2c_script
+ *
+ * Description:
+ *   Executes given script through i2c to configuure ALC1019 device.
+ *
+ ************************************************************************************/
+
+static int alc1019_exec_i2c_script(FAR struct alc1019_dev_s *priv, t_codec_init_script_entry *script, uint32_t size)
+{
+	uint32_t i;
+        uint16_t ret = 0;
+
+        for (i = 0; i < size; i++) {
+		ret = alc1019_writereg_3byte(priv, script[i].addr_h, script[i].addr_l, script[i].val);
+                if (script[i].delay > 0) {
+                	up_mdelay(script[i].delay);
+                }
+        }
+        return ret;
+}
+
+/************************************************************************************
+ * Name: alc1019_takesem
+ *
+ * Description:
+ *  Take a semaphore count, handling the nasty EINTR return if we are interrupted
+ *  by a signal.
+ *
+ ************************************************************************************/
+static void alc1019_takesem(sem_t *sem)
+{
+        int ret;
+
+        do {
+                ret = sem_wait(sem);
+                DEBUGASSERT(ret == 0 || errno == EINTR);
+        } while (ret < 0);
+}
+
+/************************************************************************************
+ * Name: alc1019_setvolume
+ *
+ * Description:
+ *   Set the right and left volume values in the ALC1019 device based on the current
+ *   volume and balance settings.
+ *
+ ************************************************************************************/
+#ifndef CONFIG_AUDIO_EXCLUDE_VOLUME
+static void alc1019_setvolume(FAR struct alc1019_dev_s *priv)
```

Comment:
please check volume level of alc1019 increase/decrease sequentially or exponentially

Reply:
For each step the volume increases by 0.375dB and it can go upto 255. Do we need to scale it again?